pageserver/src/tenant/timeline/detach_ancestor.rs

```diff
@@ -629,28 +631,39 @@ async fn copy_lsn_prefix(
     }
 }
 
-/// Creates a new Layer instance for the adopted layer, and ensures it is found from the remote
-/// storage on successful return without the adopted layer being added to `index_part.json`.
+/// Creates a new Layer instance for the adopted layer, and ensures it is found in the remote
+/// storage on successful return. without the adopted layer being added to `index_part.json`.
 async fn remote_copy(
     adopted: &Layer,
     adoptee: &Arc<Timeline>,
     generation: Generation,
     shard_identity: ShardIdentity,
     cancel: &CancellationToken,
 ) -> Result<Layer, Error> {
-    // depending if Layer::keep_resident we could hardlink
-
     let mut metadata = adopted.metadata();
     debug_assert!(metadata.generation <= generation);
     metadata.generation = generation;
     metadata.shard = shard_identity.shard_index();
 
-    let owned = crate::tenant::storage_layer::Layer::for_evicted(
-        adoptee.conf,
-        adoptee,
-        adopted.layer_desc().layer_name(),
-        metadata,
-    );
+    let conf = adoptee.conf;
+    let file_name = adopted.layer_desc().layer_name();
+
+    // depending if Layer::keep_resident, do a hardlink
+    let owned = if let Some(adopted_resident) = adopted.keep_resident().await {
+        let adopted_path = adopted_resident.local_path();
+        let adoptee_path = local_layer_path(
+            conf,
+            &adoptee.tenant_shard_id,
+            &adoptee.timeline_id,
+            &file_name,
+            &metadata.generation,
+        );
+        std::fs::hard_link(adopted_path, &adoptee_path)
+            .map_err(|e| Error::launder(e.into(), Error::Prepare))?;
```

Comment:
It would be nice for robustness in depth to rule out the hardlink replacing an existing directory entry here; like we do when finishing layer creation:
https://github.com/neondatabase/neon/commit/c508d3b5fa73d7a301f561ac8854d49ed31813be

But I guess if we're not checking the same for the remote storage either, so, probably moot to do the check locally.

Reply:
Also, technically, we should fsync the directory if we hardlinked something, as noted here:

https://github.com/neondatabase/neon/commit/3c9b484c4dbd52e286f17f3c6a5c6691990aa983

I think we don't need to because adding the hardlink is atomic.
The inode the hardlink points to is already fsynced, so, losing the hardlink would simply mean we lose the benefit of this optimization.

We can probably just remove that todo?

Reply:
added an fsync.